### ISSUE TITLE
Indexing a tensor with a boolean array raises erroneous index error only for some tensor shapes

### ISSUE THREAD

**Issue by cbirchsy** [pytorch/pytorch#121274]
### 🐛 Describe the bug

When indexing a tensor with a boolean array (numpy or default list) an erroneous index error is raised for some array shapes

```
import torch
import numpy as np

k = 31
shape = (k, 100) # shape of tensor and boolean mask
x = torch.rand(shape) # create random tensor to be indexed
mask = np.random.rand(*shape) > 0.5 # create random boolean array
assert x.shape == mask.shape # is true 

x[mask] # index the tensor
```

For $k\geq32$:
- `x[mask]`  and `x[mask.tolist()]` work as expected

For $k<32$:
- `x[mask]`  and `x[mask.tolist()]` yield ```IndexError: too many indices for tensor of dimension 2```
- For $k<3$ the error message is slightly different e.g ```IndexError: The shape of the mask [100] at index 0 does not match the shape of the indexed tensor [1, 100] at index 0```

Same pattern applies to 3D, 4D tensor etc and the shape dependence of the exception only seems to apply to zeroth dimension.

I fixed my code by using `x[torch.tensor(mask)]` which works in all cases. Nevertheless, it was tricky to catch due to the spurious and inconsistent nature of the exception raised.

### Versions

Python 3.10.12
Versions of relevant libraries:
[pip3] numpy==1.25.2
[pip3] torch==2.1.0+cu121

**Comment by ZailiWang** [comment]
Hi, I cannot reproduce the error with a recent nightly Pytorch. Would you verify again with a latest nightly build? thanks.

**Comment by cbirchsy** [comment]
Checking different versions it looks likes this bug is fixed since version 2.2.0
